### PR TITLE
Dynamic sys auth

### PR DESCRIPTION
**Pull request by talekhinezh**
Implement system authorization which can refer to component state for resource references

### CODE REVIEW

File: scrypto/src/resource/proof_rule.rs [hunk 1]
```diff
@@ -1,94 +1,139 @@
...
+
+/// Authorization Rule
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Describe, TypeId, Encode, Decode)]
+pub enum ProofRule {
+    This(ProofRuleResource),
+    AmountOf(Decimal, ProofRuleResource),
+    CountOf(u8, ProofRuleResourceList),
+    AllOf(ProofRuleResourceList),
+    AnyOf(ProofRuleResourceList),
+}
```
Comment (iamyulong) [inline scrypto/src/resource/proof_rule.rs:72]: Minor: is it possible to make the model names consistent with the macro names, unless you want to introduce another layer of abstraction?

File: radix-engine/src/model/method_authorization.rs [hunk 2]
```diff
@@ -2,105 +2,163 @@ use crate::errors::RuntimeError;
...
 #[derive(Debug, Clone, PartialEq, Eq, Hash, TypeId, Encode, Decode)]
-pub enum MethodAuthorization {
-    Protected(ProofRule),
-    Public,
-    Private,
-    Unsupported,
+pub enum HardProofRuleResource {
+    NonFungible(NonFungibleAddress),
+    Resource(ResourceDefId),
+    SoftResourceNotFound,
```
Comment (iamyulong) [inline radix-engine/src/model/method_authorization.rs:13]: Should not mix errors with model

File: radix-engine/src/model/method_authorization.rs [hunk 3]
```diff
@@ -2,105 +2,163 @@ use crate::errors::RuntimeError;
...
-                        }
-                    }
-                }
-
-                Err(NotAuthorized)
+    pub fn check_has_amount(&self, amount: Decimal, proofs_vector: &[&[Proof]]) -> bool {
+        for proofs in proofs_vector {
+            if proofs
+                .iter()
+                .any(|p| self.proof_matches(p) && p.total_amount() >= amount)
```
Comment (iamyulong) [inline radix-engine/src/model/method_authorization.rs:39]: We need to add a FIXME here.  The logic is wrong and max sum should be used.

File: radix-engine/src/model/method_authorization.rs [hunk 4]
```diff
@@ -2,105 +2,163 @@ use crate::errors::RuntimeError;
...
+impl From<ResourceDefId> for HardProofRuleResource {
+    fn from(resource_def_id: ResourceDefId) -> Self {
+        HardProofRuleResource::Resource(resource_def_id)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, TypeId, Encode, Decode)]
+pub enum HardProofRuleResourceList {
+    List(Vec<HardProofRuleResource>),
+    SoftResourceListNotFound,
```
Comment (iamyulong) [inline radix-engine/src/model/method_authorization.rs:74]: Again, mix of error code with model.

File: scrypto/src/resource/proof_rule.rs [hunk 5]
```diff
@@ -1,94 +1,139 @@
...
+
+impl From<&str> for SborPath {
+    fn from(str: &str) -> Self {
+        let path: Vec<usize> = str
+            .split('/')
+            .map(|s| s.parse::<usize>().unwrap())
+            .collect();
+        SborPath(path)
+    }
+}
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply scrypto/src/resource/proof_rule.rs:25]: good point!

File: scrypto/src/resource/proof_rule.rs [hunk 6]
```diff
@@ -1,94 +1,139 @@
...
-            ProofRule::AllOf(_) => ProofRule::OneOf(vec![self, other]),
-            ProofRule::OneOf(mut rules) => {
-                rules.push(other);
-                ProofRule::OneOf(rules)
-            }
-            ProofRule::CountOf { count: _, rules: _ } => ProofRule::OneOf(vec![self, other]),
-        }
+    Resource(ResourceDefId),
+    FromComponent(SborPath),
+}
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply scrypto/src/resource/proof_rule.rs:32]: hrm, interesting point. I was going for the more flexible, dynamic approach. Let me see what I can do here.

File: scrypto/src/resource/proof_rule.rs (hunk 1 above)
Comment (talekhinezh) [reply scrypto/src/resource/proof_rule.rs:72]: yeah, gonna keep this like this for now as still unclear number of layers of abstraction needed.

File: radix-engine/src/model/method_authorization.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine/src/model/method_authorization.rs:13]: good point!

Comment (talekhinezh) [reply radix-engine/src/model/method_authorization.rs:13]: actually, this isn't exactly an error

File: radix-engine/src/model/method_authorization.rs (hunk 3 above)
Comment (talekhinezh) [reply radix-engine/src/model/method_authorization.rs:39]: good point! though I think this is where the composition logic would be most useful.

Comment (iamyulong) [reply radix-engine/src/model/method_authorization.rs:40]: Yeah, we will need to refactor the composition logic in Proof to make it reusable.

**Review by iamyulong — APPROVED** [review]

**Pull request merged**